scalyr_agent/copying_manager/worker.py

```diff
@@ -1032,18 +1057,30 @@ def __write_checkpoint_state(
         """
         # Create the format that is expected.  An overall dict with the time when the file was written,
         # and then an entry for each file path.
-        checkpoints = {}
+        checkpoints_to_write = {}
+        all_checkpoints = {}
 
         for processor in log_processors:
+            log_path = processor.get_log_path()
+            processor_checkpoint = processor.get_checkpoint()
+
             if full_checkpoint or processor.is_active:
-                checkpoints[processor.get_log_path()] = processor.get_checkpoint()
+                checkpoints_to_write[log_path] = processor_checkpoint
 
             if full_checkpoint:
                 processor.set_inactive()
 
+            # we also save the full collection of the checkpoints in order to share this information
+            # with the copying manager.
+            all_checkpoints[log_path] = processor_checkpoint
+
+        with self._checkpoints_lock:
```

Comment:
Just thinking out loud - is there a chance for race - e.g. ``get_full_checkpoints`` is called from another thread when ``__write_checkpoint_state`` method is running, but before ``self._full_checkpoints`` is updated?

I imagine it's probably a very short time windows and likely not worth locking on the whole method level.

Reply:
I am also not 100% sure about that.
The chance of this race should be the same as the chance of the original "File is busy by another process" error on windows. So, as we could see from the original [issue](https://scalyr.myjetbrains.com/youtrack/issue/CT-274) it may happen pretty often. The main question is - what is an impact on the log reading process if the `get_full_checkpoints` has returned a bit outdated checkpoint states? I think, in theory, this may lead to the duplicated ingestion. Now I am starting to think that it may be better to wrap it all by the lock...

Reply:
@czerwingithub what do you think about that?

Reply:
I think the race really only leads to bad things in the case where we are about to start copying a log file.  We want to be sure when we start copying `file_a.log` that we are picking up from the latest place we processed.  If that's the case though, then no worker is actively processing `file_a.log` as it is.  So, if no one is currently processing it, then no one is going to be racing here, and we aren't going to get a slightly stale value for the checkpoint.

It does make me think we need to make sure we understand the different regimes here and where the latest checkpoint state for any given log file could be.  I made that point earlier about what happens when a worker has stopped processing a log file but we haven't consolidated all of the checkpoint files again.

In my mental model:

Regime 1: At the start of the agent process, all checkpoints are consolidated into a central checkpoint file.
Regime 2: A worker begins processing a log file.  At this point, the checkpoint state is stored just in the worker's checkpoint file.  More generally, I think whenever a log file has been processed by the agent, we have to assume it's checkpoint state could be in any of the worker's checkpoint files (because the file has moved between workers over time).
Regime 3: A log file has been processed by the agent, but is no longer being processed.  I think this just boils down to regime 2 again... the latest checkpoint state could be held in any of the workers checkpoint files.

All of this is making me think there are more holes in how we are handling checkpoints across the workers.  For example, what happens when a file has been processed by multiple workers and we go to consolidate all of the worker's checkpoint files?  If `file_a.log` exists in multiple worker checkpoints, there is no guarantee that the consolidated version will actually use the latest checkpoint.

Reply:
Thanks for bringing this up.

I assumed that checkpoints are also consolidated when a worker itself is stopped, but this may actually not be the case (need to dig in again - I know we do it on the whole agent process start up and shutdown).

And even if we do use this logic - there is still a chance for a race condition in case correct locking is not place.

In any case, good catch and it indeed sounds like we have more work to do. We should probably start with tests for those scenarios and then decide on the approach.

One option would be to also store timestamp in the checkpoint file (iirc, we already do that, but need to double check) and then during consolidation, also check the timestamps and use entry for a particular file with the most recent timestamp. We would of course still need to use a lock to avoid race conditions.

Reply:
> I assumed that checkpoints are also consolidated when a worker itself is stopped, but this may actually not be the case (need to dig in again - I know we do it on the whole agent process start up and shutdown).

With the current approach, I think we even have to consolidate any time a log processor is stopped on a worker, not just when an entire worker is stopped.

Yeah, we could go to per-entry timestamps.  That's where I went as well.  That could get messy and will increase the size of the checkpoint file.

The other approach I was contemplating was to just enforce a few key invariants that make reasoning about everything else easier.  For example, the workers only ever write checkpoints for files it is currently processing.  Once it stops processing a file, it will no longer include it in its checkpoint state.  The reasonability for persisting checkpoints for  files no longer being actively processed falls on the main agent process.  Of course, how we handle handing off that state in a safe manner might be tricky, but it might be worth it because it gives us easier invariants to reason with everywhere else.